### ISSUE TITLE
Return string data from a screenshot

### ISSUE THREAD

**Issue by jonnywilliamson** [nesk/puphpeteer#17]
Hi, I've just started to use puphpeteer and its working very nicely. Thank you for the library.

I do have one query regarding how to get data back from the `screenshot` method if I do NOT wish to save it to a file.

At the moment to get a screenshot your docs say:

```php
use Nesk\Puphpeteer\Puppeteer;

$puppeteer = new Puppeteer;
$browser = $puppeteer->launch();

$page = $browser->newPage();
$page->goto('https://example.com');
$page->screenshot(['path' => 'example.png']);

$browser->close();
```

Taking a look at the API docs over on the puppeteer site:

https://pptr.dev/#?product=Puppeteer&version=v1.6.2&show=api-pagescreenshotoptions

![image](https://user-images.githubusercontent.com/2513663/43960772-f98de700-9caa-11e8-9d3b-39fe8a1434cd.png)

I would like to be able to grab a base64 (or binary) string of the image received and process it directly rather than have to save it to disk.

Would there be anyway to do that at the moment, say something like this:

```php
use Nesk\Puphpeteer\Puppeteer;

$puppeteer = new Puppeteer;
$browser = $puppeteer->launch();

$page = $browser->newPage();
$page->goto('https://example.com');
$imageString = $page->screenshot(['encoding' => 'base64']);

//Do things with $imageString without having to worry about a file saved on disk

$browser->close();
```

At the moment I can't find a way to get any data out of the `BasicResource` object that gets returned from `->screenshot`

Thank you!

**Comment by quocthinh212** [comment]
Hi, I have idea.
You can save image file and get encod base64 from this file,
Delete it after you got base 64 string.

`
 $page->screenshot(['path' => 'ramdomname.png']);
            $image = base64_encode(file_get_contents("ramdomname.png"));
            unlink("ramdomname.png")
`

**Comment by jonnywilliamson** [comment]
@quocthinh212 Hi

Yeah absolutely I could do that, quick temporary file, read contents, do something, check the file is still there, delete the file if it still exists, ensure I catch any errors during that entire process etc etc.

But I was wondering and hoping I could just get the data directly from the `->screenshot` method and process it the way I would like to without all that hassle.

I'd even settle for the php://stream object that contains the data. I have had a look and it appears it's likely held in one of these streams:

![image](https://user-images.githubusercontent.com/2513663/43988513-4bcccbe2-9d2f-11e8-9bc3-6e0d24e2abb6.png)


But I have no idea at the moment the best way to get access to that, or if it's even possible.

Thank you

**Comment by quocthinh212** [comment]
could you try 
```php
$encode = $page->screenshot(['encoding' => 'base64']);
```

**Comment by jonnywilliamson** [comment]
Hi @nesk 

So, I will fill out that bug report form for you, but I've just noticed something.

When I was testing locally, I was using the binary option:

````php
    $imageString = $page->screenshot(['encoding' => 'binary']);
```

As soon as I dump that result, I get the `BasicResource` object.




But as soon as I change to 
```php
    $imageString = $page->screenshot(['encoding' => 'base64']);
```

I get the string as I wanted and as you show above!

I'm not sure why I asked for the base64 option at the start of this issue, as I do actually want the proper binary data....however, I have no idea why it works in one case and not the other.

Would you try your code again using the binary flag please?

```php
use Nesk\Puphpeteer\Puppeteer;

$puppeteer = new Puppeteer;
$browser = $puppeteer->launch();

$page = $browser->newPage();
$page->goto('https://example.com');
$imageString = $page->screenshot(['encoding' => 'binary']);

var_dump($imageString);

$browser->close();

```

**Comment by nesk** [comment]
Oh, no need to fill the template issue now, I understand what happened!

See, when you use the `binary` encoding, Puppeteer returns a `Buffer` object. However, this is a class belonging to Node and it doesn't exist in PHP, so you only get a `BasicResource` which represents the `Buffer` object (you can check this by calling `->getResourceIdentity()->className()` on the `BasicResource`).

You can perfectly call [`Buffer` methods](https://nodejs.org/api/buffer.html) on the `BasicResource,` however, I think the base64 encoding should be way easier for you to use. 😉

**Comment by jonnywilliamson** [comment]
Thanks @nesk 

Yes ok, I have no problem getting the base64 string now and converting to binary etc. Thank you for that.

However, I've now stumbled on another small issue that is probably simple to fix but I'm not sure why it's happening. Probably some sort of race condition or async problem.

If I decide to take 2 screenshots in a row, 1 screenshot never gets returned properly.

Here's the most simple code i can recreate the problem with:
```php
    $puppeteer = new Puppeteer;
    $browser = $puppeteer->launch();
    $page = $browser->newPage();
    $page->goto('https://example.com');

    $data = base64_decode($page->screenshot(['encoding' => 'base64', 'type' => 'png']));
    var_dump("image1".$data);

    $data = base64_decode($page->screenshot(['encoding' => 'base64', 'type' => 'png']));
    var_dump("image2".$data);

```


When I run this multiple times I get the following:

![image](https://user-images.githubusercontent.com/2513663/43992920-9ecc30fa-9d7d-11e8-83ca-6fd158144a14.png)


So I took a look at it using the x-debug. If I set a breakpoint AFTER the last var_dump, I see that yes indeed, only one image was returned:

![image](https://user-images.githubusercontent.com/2513663/43993058-e74baf3e-9d7f-11e8-9995-7ec6cfec7d6f.png)



If I put a breakpoint BEFORE the first screenshot and step through the code, BOTH images appear:

![image](https://user-images.githubusercontent.com/2513663/43993077-37890d20-9d80-11e8-8d43-3d47b46fc1dd.png)

It's as if the code is executing too fast in the first example, and when I slow down and step through each line slowly in the debugger, the system has a chance to get the data for me.

I don't know how else to explain it, but I was hoping you might have some further insight.

**Comment by nesk** [comment]
@jonnywilliamson Actually, I don't know what's happening here... Could you [provide me some logs](https://github.com/nesk/puphpeteer/blob/master/.github/contributing.md#provide-logs-with-your-bug-report) to help you?

**Comment by jonnywilliamson** [comment]
@nesk Thanks for getting back to me.

This is a really hard issue to replicate *reliably*. I get about 40% success and 60% failure, but I can not find out the common link between the failures.

Here are some logs and some code I'm using, (in a laravel app, in case that matters, but i'm just using a simple route closure for ease of testing).

```php
use Nesk\Puphpeteer\Puppeteer;

Route::get('/test', function () {

    $logPath = storage_path('puppeteer-example.log');
    $logger = new \Monolog\Logger('PuPHPeteer');
    $logger->pushHandler(new \Monolog\Handler\StreamHandler($logPath, \Monolog\Logger::DEBUG));
    $puppeteer = new Puppeteer(['logger' => $logger]);

    $browser = $puppeteer->launch();
    $page = $browser->newPage();
    $page->goto('https://example.com');

    $image1 = $page->screenshot(['encoding' => 'base64', 'type' => 'png']);
    dump("The data from image1 is: ".$image1);
    $image2 = $page->screenshot(['encoding' => 'base64', 'type' => 'png']);
    dump("The data from image2 is: ".$image2);
});
```

`dump()` is part of  symfony that makes output to the screen easy to see/formatted. Using `var_dump` etc will work obviously just as well.

The URL's I've used in this test are
`https://example.com`
`https://google.com`
`https://facebook.com`
`https://yahoo.com`

Both example.com and google.com WORKED perfectly. I got 2 image strings dumped.
Both facebook and yahoo DID NOT work, neither string got dumped.

What is more annoying is that yesterday, both google.com and others would only dump ONE string, and it was random if it was going to be the first or second one. I couldn't detect a pattern. Today, they dumped BOTH strings! Grrrr...

In almost all occassions, if I turn on x-debug, and hit a breakpoint in the code and step through it, both strings appear to get dumped. I'm very confused as to where the issue is - my code, node server, or puphpeteer.

Here are the logs:
[puppeteer-example.log](https://github.com/nesk/puphpeteer/files/2297164/puppeteer-example.log)
[puppeteer-facebook.log](https://github.com/nesk/puphpeteer/files/2297165/puppeteer-facebook.log)
[puppeteer-google.log](https://github.com/nesk/puphpeteer/files/2297166/puppeteer-google.log)
[puppeteer-yahoo.log](https://github.com/nesk/puphpeteer/files/2297167/puppeteer-yahoo.log)

I hope you can recreate this on your end.

Many thanks for replying.

**Comment by jonnywilliamson** [comment]
One last item that seems to add to the mystery on my machine. If I reset my computer, I can get the code above to work the first time it's run reliably.....but then if run a second or more times it is no longer guaranteed to work!

I'm beat!

**Comment by nesk** [comment]
Okay, I will try to be more responsive the next time, I just tried your code (multiple times) and had no issues. However, I checked your log files, looked once again at your previous messages, and there's something weird: you're asking for a base64 representation of your screenshots but sometimes you get some binary code instead of base64.

So, as I said before:

> To help you I will need you to post a new message with all the informations asked [in the "Bug report" template](https://github.com/nesk/puphpeteer/blob/master/.github/ISSUE_TEMPLATE/bug_report.md).

Also, what version of PuPHPeteer are you running? Also, are you running your PHP code with Apache or Nginx (specify the version for both)?

**Comment by nesk** [comment]
Did you try to reproduce this bug by using Puppeteer without PuPHPeteer? If it fails also in pure JS, you will have to report this in [GoogleChrome/puppeteer](https://github.com/GoogleChrome/puppeteer) issues.

**Comment by jonnywilliamson** [comment]
**Describe the bug**

Whilst attempting to take two screenshots of a page consecutively and returning each screenshot as a base64 string (rather than saving as a file), not every screenshot string gets returned successfully.

In some cases all work, in others only 1 works, in some cases none work.

**Reproducible example**
```php
use Nesk\Puphpeteer\Puppeteer;

Route::get('/test', function () {

    $logPath = storage_path('puppeteer-example.log');
    $logger = new \Monolog\Logger('PuPHPeteer');
    $logger->pushHandler(new \Monolog\Handler\StreamHandler($logPath, \Monolog\Logger::DEBUG));
    $puppeteer = new Puppeteer(['logger' => $logger]);

    $browser = $puppeteer->launch();
    $page = $browser->newPage();
    $page->goto('https://example.com');

    $image1 = $page->screenshot(['encoding' => 'base64', 'type' => 'png']);
    dump("The data from image1 is: ".$image1);
    $image2 = $page->screenshot(['encoding' => 'base64', 'type' => 'png']);
    dump("The data from image2 is: ".$image2);

});
```

**Expected behavior**
Taking multiple screenshots and returning the data as a base64 string should be successful every time OR throw an exception if there was an error in processing.

**Generated logs**

[puppeteer-example.log](https://github.com/nesk/puphpeteer/files/2297164/puppeteer-example.log)
[puppeteer-facebook.log](https://github.com/nesk/puphpeteer/files/2297165/puppeteer-facebook.log)
[puppeteer-google.log](https://github.com/nesk/puphpeteer/files/2297166/puppeteer-google.log)
[puppeteer-yahoo.log](https://github.com/nesk/puphpeteer/files/2297167/puppeteer-yahoo.log)


**Screenshots**
If applicable, add screenshots to help explain your problem.

**Environment (please complete the following information):**
 - OS: [e.g. Windows 10, macOS 10.13.2, Ubuntu 18.04]
Laravel Homestead running on a Windows 10 Professional computer. 
The Ubuntu version that homestead is running is: Ubuntu 16.04.4 LTS

 - Node version: [e.g. 10.4.0]
v8.11.2

 - PHP version [e.g. 7.2.6]
PHP 7.2.3-1+ubuntu16.04.1+deb.sury.org+1

My Node package manager is:
 - [ ] Yarn (specify the version)
Yarn 1.7.0

Puphpeteer Version:
nesk/puphpeteer                       1.2.0 

Running on Nginx
nginx version: nginx/1.13.6


**Additional context**
Having had a look at the log files, I note as you did that whilst the string data was passed back to puphpeteer after the screenshot was taken, on some results at the end of the string binary data appears.

Here's a small example
![image](https://user-images.githubusercontent.com/2513663/44530202-cb19fe00-a6e5-11e8-8a64-2c679c0c40cb.png)


Looking at some of those bytes they remind me of TCP/IP responses/packets that have been not been processed?

Eg 
NAK - negative acknowledgment
SYN - synchronize

etc. Perhaps there really is an issue further up the chain.



I'm also just going to be honest and say I don't know how I would test the same code using JS only to see if it was a puppeteer issue instead. 😞

**Comment by jonnywilliamson** [comment]
Ok I found out how to run with plain JS:

```js

const puppeteer = require('puppeteer');

(async () => {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();
  await page.goto('https://example.com');

  let one = await page.screenshot({
      encoding: 'base64'
  });

    let two = await page.screenshot({
        encoding: 'base64'
    });

    console.log('one'+one+"\n\n\n");
    console.log('two'+two+"\n\n\n");
    await browser.close();
})();
```

I have run this multiple times, with different URLs and different number of screenshots. On the same computer/server as the php code and the screenshot string gets returned every time.

I went back and did a test run again with the php code calling puphpeteer and immediately 1 of the screenshots did not complete out of the 2. 😢 

![image](https://user-images.githubusercontent.com/2513663/44531978-f3a3f700-a6e9-11e8-9047-ffa5ba31001f.png)

[puppeteer-example.log](https://github.com/nesk/puphpeteer/files/2314903/puppeteer-example.log)

**Comment by jonnywilliamson** [comment]
And now I see you have a new version released 3 days ago. More importantly, I see that nesk/rialto is a version 1.2.0

I have been using nesk/rialto 1.1.0 and have been playing in the `ProcessSupervisor.php` class where I can see the data coming into the system. Some of the chunks in the `readNextProcessValue` where not getting base64_decoded properly....

https://github.com/nesk/rialto/blob/1.1.0/src/ProcessSupervisor.php#L311-L336


Going to upgrade and see if that helps.

**Comment by nesk** [comment]
Homestead seems to be the cause, I'm reproducing your issue. I will try to fix this.

**Comment by jonnywilliamson** [comment]
Thank you for the update. 

Whilst I'm not happy you've found the same issue and have to spend valuable time debugging it,  I am very happy that you've confirmed I'm not completely crazy! 😂

**Comment by nesk** [comment]
So, I _might_ have a solution. It works for me in Laravel Homestead and I really hope it will work for you as well.

Go in your `composer.json` file and add this to your requirements:

```js
"require": {
    // your other requirements...

    "nesk/rialto": "dev-next-chunk-delay as 1.2.1",
}
```

Run `composer update nesk/rialto` and try to run once again the script which was failing, it should work now.

If it works, let me know and I will make a new release for Rialto based on this fix. This way you will be able to remove this nasty requirement (and I will be able to delete the `next-chunk-delay` branch).

<hr>

>Whilst I'm not happy you've found the same issue and have to spend valuable time debugging it,  I am very happy that you've confirmed I'm not completely crazy! 😂

Well, if you find my time valuable enough, [feel free to buy me a beer 🍺](https://github.com/nesk/puphpeteer#support-me) once this issue is solved. 😉

**Comment by jonnywilliamson** [comment]
Hi @nesk 

Brilliant. I've just tested this out and it does indeed fix the issue! I am getting ALL responses back successfully.

I have thoughts about the [commit you made](https://github.com/nesk/rialto/commit/2dedb65ac520971c9e47a1724f98632617c65bf9) on nesk/rialto and so I'm going to continue the conversation over there. Of course I owe you a beer :)

**Comment by nesk** [comment]
I will make a new release based on this code on Tuesday. But in the future a refactoring will be done to clean the socket part of my code, to avoid these crappy workarounds. 😉

**Comment by nesk** [comment]
@jonnywilliamson I've just released Rialto's v1.2.1. You only have to remove the `nesk/rialto` requirement in your `composer.json` file and run `composer update nesk/rialto`. 🙂
